### PR TITLE
feat: add multi-provider to SDK

### PR DESCRIPTION
**Pull request by Vishalup29**
## Description

Move the existing multi-provider implementation from the contrib module into the core Java SDK, mark it as experimental, and deprecate the contrib implementation.

## Changes

- Move `MultiProvider` (and related strategy/utility classes) from `java-sdk-contrib/providers/multiprovider` into the core SDK under `dev.openfeature.sdk.providers.multiprovider`.
- Mark the SDK multi-provider as **experimental** using inline Javadoc and README documentation.
- Add SDK tests for the multi-provider based on the existing contrib tests.
- Update the Java SDK `README.md` to document the multi-provider feature and show a basic usage example.
- Deprecate the contrib multi-provider implementation and README, pointing users to the SDK implementation.
- Adjust exception usage to match the SDK API (use `new FlagNotFoundError()` rather than a message constructor).

## Related Issues

Fixes open-feature#1486

## Notes

- In the SDK, `FlagNotFoundError` only exposes a no-arg constructor, so the ported multi-provider now throws `new FlagNotFoundError()` where the contrib version used a message.
- Metadata handling has been aligned with the SDK’s existing patterns and dependency constraints to avoid introducing unnecessary dependencies.

### CODE REVIEW

**Comment by Vishalup29** [comment]
@chrfwow @toddbaert Can you pls review when you get a chance?

**Comment by aepfli** [comment]
Hi, thanks for your contribution and interest in the SDK! We genuinely appreciate your work on multi-provider support.

Since there's already a well-developed PR (#1500 ) addressing the same feature, having multiple PRs for the same goal creates extra effort for reviewers and can slow things down. Next time, we suggest building on the existing work - especially when significant development and review have already occurred.

A few differences to keep in mind:
- PR #1500 offers comprehensive tests, standardized exception handling, and has removed unnecessary dependencies.
- It includes extensive review feedback covering several improvements and edge cases.

Your contributions are always welcome, but i think basing the work on the previous PR would be a good approach here. Especially to also credit the previous contributor of his base work.

Thank you for your effort, can you maybe incorporate the already existing feedback from the other pull request, or re-base your efforts based on the other one. This would be highly appreciated.

**Comment by Vishalup29** [comment]
@aepfli Rebased and PR on top of the changes in the PR (https://github.com/open-feature/java-sdk/pull/1500 ). it now includes all the suggestions based on the discussions on that PR. Pls take a look when you get a chance

File: src/test/java/dev/openfeature/sdk/multiprovider/BaseStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
+package dev.openfeature.sdk.multiProvider;
```
[the remark this answers is not in the document]
Comment (aepfli) [reply src/test/java/dev/openfeature/sdk/multiprovider/BaseStrategyTest.java]: seems like a valid reasoning, and we should rename the pacakge

[the remark this answers is not in the document]
Comment (Vishalup29) [reply src/test/java/dev/openfeature/sdk/multiprovider/BaseStrategyTest.java]: addressed

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     * @param providers the list of FeatureProviders to initialize the MultiProvider with
+     * @param strategy  the strategy (if {@code null}, {@link FirstMatchStrategy} is used)
+     */
+    public MultiProvider(List<FeatureProvider> providers, Strategy strategy) {
+        this.providers = buildProviders(providers);
+        if (strategy != null) {
+            this.strategy = strategy;
+        } else {
+            this.strategy = new FirstMatchStrategy();
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: Should we assume that `null` means the user wants to have a `FirstMatchStrategy`?

I'd rather prefer a `NullReferenceException` here and only use the `FirstMatchStrategy` in the single parameter `MultiProvider(List<FeatureProvider> providers)` constructor.
WDYT?

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        }
+    }
+
+    protected static Map<String, FeatureProvider> buildProviders(List<FeatureProvider> providers) {
+        Map<String, FeatureProvider> providersMap = new LinkedHashMap<>(providers.size());
+        for (FeatureProvider provider : providers) {
+            FeatureProvider prevProvider =
+                    providersMap.put(provider.getMetadata().getName(), provider);
+            if (prevProvider != null) {
+                log.warn("duplicated provider name: {}", provider.getMetadata().getName());
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: Why do we want to prevent this? @aepfli, @chrfwow do we treat the provider name as unique identifier in the SDK?

And if, we should throw because it would be a clear misconfiguration.

Scenario where I could see the same provider twice is e.g., using flagd providers that should read from multiple flag configuration files. According to the spec, the [provider name](https://openfeature.dev/specification/sections/providers#requirement-211) identifies the implementation, not the instance.

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+ * <b>Experimental:</b> Provider implementation for multi-provider.
+ *
+ * <p>This provider delegates flag evaluations to multiple underlying providers using a configurable
+ * {@link Strategy}. It also exposes combined metadata containing the original metadata of each
+ * underlying provider.
+ */
+@Slf4j
+public class MultiProvider extends EventProvider {
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:33]: The `MultiProvider` does not emit any [lifecycle events](https://openfeature.dev/specification/sections/events). If we don't want to support this in the first iteration it is sufficient to implement `FeatureProvider` instead of extending `EventProvider`.

Although we should take care of this in a follow up asap, because we effectively loose the Event features when using `MultiProvider`.

File: src/main/java/dev/openfeature/sdk/multiprovider/FirstMatchStrategy.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    // Successful evaluation
+                    return res;
+                }
+                if (!FLAG_NOT_FOUND.equals(errorCode)) {
+                    // Any non-FLAG_NOT_FOUND error bubbles up
+                    return res;
+                }
+                // else FLAG_NOT_FOUND: skip to next provider
+            } catch (FlagNotFoundError e) {
+                log.debug(
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/multiprovider/FirstMatchStrategy.java]: We usually do not log in hot code paths like flag evaluations.

File: src/main/java/dev/openfeature/sdk/multiprovider/FirstSuccessfulStrategy.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            Function<FeatureProvider, ProviderEvaluation<T>> providerFunction) {
+        for (FeatureProvider provider : providers.values()) {
+            try {
+                ProviderEvaluation<T> res = providerFunction.apply(provider);
+                if (res.getErrorCode() == null) {
+                    // First successful result (no error code)
+                    return res;
+                }
+            } catch (Exception e) {
+                log.debug(
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/multiprovider/FirstSuccessfulStrategy.java]: We usually do not log in hot code paths but can see the value of having a debug message here.

**Review by guidobrei — COMMENTED** [review]
I've gone over the Implementation once more and added some concerns to be discussed.
It does not block this PR which is about moving the provider from contrib to the SDK and afaik the issues currently exist in the contrib implementation as well. But maybe for the valid points identified we could craft some new issues.

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 2 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: Thanks!! Good point, using null as a signal for the default strategy is confusing.
I’ve updated the constructors so that:

MultiProvider(List<FeatureProvider> providers) uses FirstMatchStrategy by default

MultiProvider(List<FeatureProvider> providers, Strategy strategy) now requires a non-null strategy and throws if it’s null
This keeps the simple ctor ergonomic while making the 2-arg ctor semantics explicit.

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 3 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: That’s a really good question, and I agree it needs a clear spec-level answer.
For this PR I’ve kept the contrib behaviour (log a warning and let the last provider win) to avoid changing semantics while we move the implementation into the SDK.
I can open a follow-up issue to clarify whether provider names should be treated as unique identifiers or implementation identifiers, and we can update the behaviour there once we agree on the desired contract.

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 4 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:33]: Agreed that the current MultiProvider doesn’t emit events yet.
For now I’ve kept it extending EventProvider to avoid changing the surface again during the move, and to keep the door open for proper event propagation in a follow-up.
I’ll open a follow-up issue to track “Emit provider lifecycle/configuration events from MultiProvider” so we can design the right event semantics without blocking this migration PR.

File: src/main/java/dev/openfeature/sdk/multiprovider/FirstMatchStrategy.java (hunk 5 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/FirstMatchStrategy.java]: Makes sense – logging inside the hot evaluation path isn’t aligned with how the rest of the SDK behaves.
I’ve removed the debug logging from both strategy implementations and now just skip on FLAG_NOT_FOUND / exceptions, relying on the individual providers (or hooks/logging hook) for diagnostics.

File: src/main/java/dev/openfeature/sdk/multiprovider/FirstSuccessfulStrategy.java (hunk 6 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/FirstSuccessfulStrategy.java]: Makes sense – logging inside the hot evaluation path isn’t aligned with how the rest of the SDK behaves.
I’ve removed the debug logging from both strategy implementations and now just skip on FLAG_NOT_FOUND / exceptions, relying on the individual providers (or hooks/logging hook) for diagnostics.

**Comment by Vishalup29** [comment]
@chrfwow @toddbaert  @aepfli @guidobrei can this be approved/merged?

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 3 above)
Comment (chrfwow) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: I don't think we want the names to be unique, but maybe it doesn't make sense to add the same provider twice to the multiprovider, and if we add telemetry which provider responded, this would make it unclear. But I believe an info log would be enough

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Getter
+    private static final String NAME = "multiprovider";
+
+    public static final int INIT_THREADS_COUNT = 8;
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: This seems a bit like an arbitrary number, maybe use `Runtime.getRuntime().availableProcessors();` instead, if it doesn't have some weird edge cases

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+     * Initialize the provider.
+     *
+     * @param evaluationContext evaluation context
+     * @throws Exception on error (e.g. wrapped {@link java.util.concurrent.ExecutionException}
+     *                   from a failing provider)
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        var metadataBuilder = MultiProviderMetadata.builder().name(NAME);
+        HashMap<String, Metadata> providersMetadata = new HashMap<>();
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:88]: Considering https://github.com/open-feature/java-sdk/pull/1765/changes#r2618228793, this should maybe be a KeyValuePair List

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            for (FeatureProvider provider : providers.values()) {
+                try {
+                    provider.shutdown();
+                } catch (Exception shutdownEx) {
+                    log.error(
+                            "error shutting down provider {} after failed initialize",
+                            provider.getMetadata().getName(),
+                            shutdownEx);
+                }
+            }
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: We could call `shutdown` here

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 7 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java]: Switched INIT_THREADS_COUNT to Runtime.getRuntime().availableProcessors() and still cap it with providers.size() when creating the thread pool.

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 8 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:88]: For now I kept originalMetadata as Map<String, Metadata> to match how buildProviders currently treats provider names as keys.
I agree a List<KeyValuePair<String, Metadata>> could better represent duplicate names; happy to open a follow-up issue to discuss that representation and adjust the metadata type there if we decide to support duplicates fully.

**Comment by Vishalup29** [comment]
@chrfwow @toddbaert @aepfli @guidobrei can this be approved/merged?

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 8 above)
Comment (chrfwow) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:88]: Have you created the issue? I think this is something we should not forget

**Comment by chrfwow** [comment]
@Vishalup29 you need to run the spotless plugin once more to fix the code format issues

File: src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java (hunk 8 above)
Comment (Vishalup29) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:88]: fixed all the issues. CI green

Comment (toddbaert) [reply src/main/java/dev/openfeature/sdk/multiprovider/MultiProvider.java:88]: I think @chrfwow meant to open a new issue for the commend here. 

I've done this: https://github.com/open-feature/java-sdk/issues/1792

**Review by toddbaert — APPROVED** [review]
I've rebased, squashed, and added @suvaidkhan, the author of #1500 , as a co-author here, since some of their work went into this.

I'm approving, but I would love one more approval from @aepfli or @guidobrei

**Comment by Vishalup29** [comment]
@chrfwow @toddbaert @aepfli @guidobrei - what is the process to get a PR merged in this repo? This PR has undergone multiple rounds of review and has also been approved multiple times. Who can merge this?

**Comment by aepfli** [comment]
I am sorry, I have been out of service since the middle of December, and I will be back next Monday. - first sick and afterwards holiday season. 

Sorry for the delays on my side. But this time of the year is usually more focused on time off the computer or other side projects, rather than my usual dev life ;)

**Comment by toddbaert** [comment]
> @chrfwow @toddbaert @aepfli @guidobrei - what is the process to get a PR merged in this repo? This PR has undergone multiple rounds of review and has also been approved multiple times. Who can merge this?

Hey @Vishalup29 ... I can merge it, but this took a while for a few reasons:

- Christmas means fewer people are working, so things are slower than usual.
- This is a 1000 line PR, and it's going into the main SDK, so thorough review is required.
- There's so-far unresolved comments from previous reviewers, so I would prefer to see them resolved by those people. If they can't soon, I will re-review myself once more, and merge.

As you can see, 5 days ago I made some changes on here myself, so don't worry, your efforts are not being ignored. We are just being thorough. :)

**Comment by toddbaert** [comment]
@Vishalup29 I did a final re-review. Things look good. Thanks so much for your contribution and patience. Sorry that we were slow on this one. Happy new year. :christmas_tree:

**Comment by Vishalup29** [comment]
Thanks @toddbaert and a very Happy New Year!
